### PR TITLE
Feat add End Meeting for All

### PR DESCRIPTION
**Pull request by sumitkumar121**
# Description #
The scope of the changes in this pull request are to add the ability to end meeting for all the participant without impacting the existing functionality

## Changes include:

1. Adding new function endmeetingforall in @webex\plugin-meetings\src\meeting\index.js
2. Created a new api request for making call to locus endpoint.
3. Updated Metrics for tracking endMettingForall.

### Change Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios where tested

Tested Manually and with Automated script that all functionality and new functionality is working.

### I certified that

- [X] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [X] I discussed changes with code owners prior to submitting this pull request

- [x] I have not skipped any automated checks
- [x] All existing and new tests passed
- [x ] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

**Comment by arun3528** [comment]
@sumitkumar121  any update on this ?

**Comment by sumitkumar121** [comment]
> @sumitkumar121 any update on this ?

This pr is ready for review

**Comment by arun3528** [comment]
hey @sumitkumar121  can we have a call to go over the changes if that ok for you ?

File: packages/node_modules/@webex/plugin-meetings/src/meeting/request.js [hunk 1]
```diff
@@ -647,4 +648,23 @@ export default class MeetingRequest extends StatelessWebexPlugin {
...
+   * @param {Object} options
+   * @param {Url} options.locusUrl
+   * @param {String} options.uuid
+   * @returns {Promise}
+   */
+  endMeetingForAll({
+    locusUrl,
+    uuid
+  }) {
+    const uri = `${locusUrl}/${uuid}/${END}`;
```
Comment (marcin-bazyl) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/request.js:663]: this doesn't look right, surely the locusUrl already contains the id to identify the meeting so appending `/${uuid}/` shouldn't be needed here

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 2]
```diff
@@ -5621,4 +5621,92 @@ export default class Meeting extends StatelessWebexPlugin {
 
     return (start && end) ? end - start : undefined;
   }
+
+
+  /**
+   * End the current meeting for all
+   * @param {Object} options leave options
+   * @param {String} options.uuid the metting id
```
[the remark this answers is not in the document]
Comment (marcin-bazyl) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:5629]: also, why is this needed at all? the endMeetingForAll() method on the meeting object should not need any arguments. SDK already knows all the information about the meeting to be able to end it.

File: packages/node_modules/@webex/plugin-meetings/src/meeting/state.js [hunk 3]
```diff
@@ -71,6 +71,15 @@ const MeetingStateMachine = {
           ],
           to: MEETING_STATE_MACHINE.STATES.ENDED
         },
+        //  when ending a meeting it must be from  JOINED, or ERROR states, and transitions it to the ENDED state
+        {
+          name: MEETING_STATE_MACHINE.TRANSITIONS.END,
+          from: [
```
Comment (marcin-bazyl) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/state.js:77]: The LEAVE transition is defined also for RINGING/ANSWERED/DECLINED states - I guess these are used only in 1-1 calls. I'm not sure are we allowing `endMeetingForAll()` to be called for 1-1 calls? If no, then I don't see any code catching that case and throwing appropriate error. If yes, then theses states should be added here.

File: packages/node_modules/@webex/plugin-meetings/src/meeting/util.js [hunk 4]
```diff
@@ -744,4 +744,40 @@ MeetingUtil.handleDeviceLogging = (devices = []) => {
   });
 };
 
+MeetingUtil.endMeetingForAll = (meeting, options = {}) => {
+  if (meeting.meetingState === FULL_STATE.INACTIVE) {
+    // TODO: clean up if the meeting is already inactive
```
Comment (marcin-bazyl) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/util.js:749]: we shouldn't have any todo comments in the code - if this needs to be done, it should be done in this PR or if for some good reason cannot be done now, raise a jira for tracking. (I know this is a copy&paste from the leave() method, but we should still try to do something about it as we're touching this area)

File: packages/node_modules/@webex/plugin-meetings/src/meeting/util.js [hunk 5]
```diff
@@ -744,4 +744,40 @@ MeetingUtil.handleDeviceLogging = (devices = []) => {
...
+
+  const defaultOptions = {
+    locusUrl: meeting.locusUrl
+  };
+
+  const leaveOptions = {...defaultOptions, ...options};
+
+  return meeting.meetingRequest
+    .endMeetingForAll(leaveOptions)
+    .then((response) => Promise.resolve(response))
```
Comment (marcin-bazyl) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/util.js:761]: this line is useless - doesn't do anything. The leave() method does a call to `onFullLocus()` here. Why are we not doing it here too?

File: packages/node_modules/@webex/plugin-meetings/src/meeting/util.js [hunk 6]
```diff
@@ -744,4 +744,40 @@ MeetingUtil.handleDeviceLogging = (devices = []) => {
...
+    .then(() => {
+      if (options.moveMeeting) {
```
Comment (marcin-bazyl) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/util.js:763]: does it make sense to have a `moveMeeting` option for "end meeting for all"?

File: packages/node_modules/@webex/plugin-meetings/src/meeting/util.js [hunk 7]
```diff
@@ -744,4 +744,40 @@ MeetingUtil.handleDeviceLogging = (devices = []) => {
...
+      }
+
+      return MeetingUtil.cleanUp(meeting);
+    })
+    .catch((err) => {
+      // TODO: If the meeting state comes as LEFT or INACTIVE as response then
+      // 1)  on leave clean up the meeting or simply do a sync on the meeting
+      // 2) If the error says meeting is inactive then destroy the meeting object
+      LoggerProxy.logger.error(
+        `Meeting:util# --> An error occured while trying to end meeting for all with an id of ${
```
Comment (marcin-bazyl) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/util.js:774]: `Meeting:util#` -> `Meeting:util#endMeetingForAll`

File: packages/node_modules/@webex/plugin-meetings/test/unit/spec/meeting/request.js [hunk 8]
```diff
@@ -249,5 +249,21 @@ describe('plugin-meetings', () => {
         assert.equal(requestParams.body.device.deviceType, 'PROVISIONAL');
       });
     });
+
+    describe('#joinMeeting', () => {
```
Comment (marcin-bazyl) [inline packages/node_modules/@webex/plugin-meetings/test/unit/spec/meeting/request.js:253]: wrong title

File: packages/node_modules/@webex/plugin-meetings/test/unit/spec/meeting/request.js [hunk 9]
```diff
@@ -249,5 +249,21 @@ describe('plugin-meetings', () => {
...
+      it('sends  requets for join', async () => {
```
Comment (marcin-bazyl) [inline packages/node_modules/@webex/plugin-meetings/test/unit/spec/meeting/request.js:254]: wrong title and contains typos

**Review by marcin-bazyl — CHANGES_REQUESTED** [review]

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 10]
```diff
@@ -5621,4 +5621,92 @@ export default class Meeting extends StatelessWebexPlugin {
...
+          this,
+          {
+            file: 'meeting/index',
+            function: 'endMeetingForAll'
+          },
+          EVENTS.REQUEST_UPLOAD_LOGS,
+          this
+        );
+        Metrics.sendBehavioralMetric(
+          BEHAVIORAL_METRICS.MEETING_LEAVE_FAILURE,
```
[the remark this answers is not in the document]
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:5699]: +1

[the remark this answers is not in the document]
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:5699]: Can you also add one for initiated in the starting

File: packages/node_modules/@webex/plugin-meetings/src/meeting/state.js [hunk 11]
```diff
@@ -71,6 +71,15 @@ const MeetingStateMachine = {
           ],
           to: MEETING_STATE_MACHINE.STATES.ENDED
         },
+        //  when ending a meeting it must be from  JOINED, or ERROR states, and transitions it to the ENDED state
+        {
+          name: MEETING_STATE_MACHINE.TRANSITIONS.END,
```
[the remark this answers is not in the document]
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/state.js:76]: The meeting will get destroyed so we wont need this i guess @marcin-bazyl

File: packages/node_modules/@webex/plugin-meetings/src/meeting/util.js (hunk 5 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/util.js:761]: +1

File: packages/node_modules/@webex/plugin-meetings/src/metrics/config.js [hunk 12]
```diff
@@ -135,6 +135,7 @@ export const eventType = {
   MUTED: 'client.muted',
   UNMUTED: 'client.unmuted',
   LEAVE: 'client.call.leave',
+  END: 'client.call.end',
```
Comment (arun3528) [inline packages/node_modules/@webex/plugin-meetings/src/metrics/config.js:138]: Previously was end event not being sent on leave ?

File: packages/node_modules/@webex/plugin-meetings/test/unit/spec/meeting/index.js [hunk 13]
```diff
@@ -2400,6 +2400,64 @@ describe('plugin-meetings', () => {
...
+        afterEach(() => {
+          sandbox.restore();
+          sandbox = null;
+        });
+        it('should end the meeting for all and return promise', async () => {
+          const endMeetingForAll = meeting.endMeetingForAll();
+
+          assert.exists(endMeetingForAll.then);
+          await endMeetingForAll;
+          assert.calledOnce(meeting.meetingRequest.endMeetingForAll);
```
Comment (arun3528) [inline packages/node_modules/@webex/plugin-meetings/test/unit/spec/meeting/index.js:2449]: add parameter check here

**Review by arun3528 — CHANGES_REQUESTED** [review]

File: packages/node_modules/@webex/plugin-meetings/src/constants.js [hunk 14]
```diff
@@ -845,6 +846,7 @@ export const MEETING_STATE_MACHINE = {
     JOIN: 'join',
     DECLINE: 'decline',
     LEAVE: 'leave',
+    END: 'END',
```
[the remark this answers is not in the document]
Comment (sumitkumar121) [reply packages/node_modules/@webex/plugin-meetings/src/constants.js]: aded in lower case

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (sumitkumar121) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: Fixed the type and removed argument. As suggested

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 15]
```diff
@@ -5621,4 +5621,92 @@ export default class Meeting extends StatelessWebexPlugin {
...
+   * @param {Object} options leave options
+   * @param {String} options.uuid the metting id
+   * @returns {Promise}
+   * @public
+   * @memberof Meeting
+   */
+  endMeetingForAll(options = {}) {
+    Metrics.postEvent({event: eventType.END, meeting: this, data: {trigger: trigger.USER_INTERACTION, canProceed: false}});
+
+    LoggerProxy.logger.log('Meeting:index#end --> End meeting for All');
```
[the remark this answers is not in the document]
Comment (sumitkumar121) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: Method name changed

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 16]
```diff
@@ -5621,4 +5621,92 @@ export default class Meeting extends StatelessWebexPlugin {
...
+
+  /**
+   * End the current meeting for all
+   * @param {Object} options leave options
+   * @param {String} options.uuid the metting id
+   * @returns {Promise}
+   * @public
+   * @memberof Meeting
+   */
+  endMeetingForAll(options = {}) {
```
[the remark this answers is not in the document]
Comment (sumitkumar121) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: Moved common code to different function

File: packages/node_modules/@webex/plugin-meetings/src/meeting/util.js (hunk 5 above)
Comment (sumitkumar121) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/util.js]: Removed the line and added onFullLocus()

File: packages/node_modules/@webex/plugin-meetings/src/meeting/util.js (hunk 6 above)
Comment (sumitkumar121) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/util.js]: Removed moveMetting from here

File: packages/node_modules/@webex/plugin-meetings/src/meeting/util.js (hunk 7 above)
Comment (sumitkumar121) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/util.js:769]: Title Name fixed

File: packages/node_modules/@webex/plugin-meetings/test/unit/spec/meeting/request.js (hunk 8 above)
Comment (sumitkumar121) [reply packages/node_modules/@webex/plugin-meetings/test/unit/spec/meeting/request.js]: Title name fixed

File: packages/node_modules/@webex/plugin-meetings/test/unit/spec/meeting/request.js (hunk 9 above)
Comment (sumitkumar121) [reply packages/node_modules/@webex/plugin-meetings/test/unit/spec/meeting/request.js]: title name and typo fixed

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 17]
```diff
@@ -5621,4 +5621,92 @@ export default class Meeting extends StatelessWebexPlugin {
...
+    Metrics.postEvent({event: eventType.END, meeting: this, data: {trigger: trigger.USER_INTERACTION, canProceed: false}});
```
[the remark this answers is not in the document]
Comment (sumitkumar121) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:5635]: we need to send same event for both leave and end confirmed by Meera Krishnaswami

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 18]
```diff
@@ -5621,4 +5621,92 @@ export default class Meeting extends StatelessWebexPlugin {
...
+          {
+            file: 'meeting/index',
+            function: 'endMeetingForAll'
+          },
+          EVENTS.REQUEST_UPLOAD_LOGS,
+          this
+        );
+
+        // TODO: more testing before we remove this code, we are not sure the scenarios for destroy here
+        if (this.wirelessShare) {
```
[the remark this answers is not in the document]
Comment (sumitkumar121) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: this code has been removed

File: packages/node_modules/@webex/plugin-meetings/src/metrics/config.js (hunk 12 above)
Comment (sumitkumar121) [reply packages/node_modules/@webex/plugin-meetings/src/metrics/config.js:138]: this is new event i have created.

File: packages/node_modules/@webex/plugin-meetings/test/unit/spec/meeting/request.js [hunk 19]
```diff
@@ -249,5 +249,20 @@ describe('plugin-meetings', () => {
         assert.equal(requestParams.body.device.deviceType, 'PROVISIONAL');
       });
     });
+
+    describe('#endmeetingForall', () => {
+      it('sends  requets to endmeetingForall', async () => {
```
[the remark this answers is not in the document]
Comment (sumitkumar121) [reply packages/node_modules/@webex/plugin-meetings/test/unit/spec/meeting/request.js]: all typo and space issue fixed.

File: packages/node_modules/@webex/plugin-meetings/src/meeting/request.js [hunk 20]
```diff
@@ -647,4 +648,22 @@ export default class MeetingRequest extends StatelessWebexPlugin {
       }
     });
   }
+
+  /**
+   * Make a network request to end meeting for all
+   * @param {Object} options
+   * @param {Url} options.locusUrl
+   * @param {String} options.lid
```
[the remark this answers is not in the document]
Comment (sumitkumar121) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/request.js]: options.lid  comment has been removed

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 21]
```diff
@@ -5635,4 +5622,89 @@ export default class Meeting extends StatelessWebexPlugin {
...
+  }
+
+   /**
+   * clear the metting data
+   * @returns {undefined}
+   * @public
+   * @memberof Meeting
+   */
+   clearMeetingData = () => {
+     this.meetingFiniteStateMachine.end();
```
[the remark this answers is not in the document]
Comment (sumitkumar121) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: Yes it is wrong leave should have this.meetingFiniteStateMachine.leave(); and end meeting should have this.meetingFiniteStateMachine.end(). Corrected this mistake.

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 22]
```diff
@@ -5635,4 +5622,89 @@ export default class Meeting extends StatelessWebexPlugin {
...
+            code: error.code
+          }
+        );
+
+        return Promise.reject(error);
+      });
+  }
+
+   /**
+   * clear the metting data
```
[the remark this answers is not in the document]
Comment (sumitkumar121) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: typo fixed

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (sumitkumar121) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: Added metrics for initiation also

File: packages/node_modules/@webex/plugin-meetings/src/meeting/state.js [hunk 23]
```diff
@@ -71,6 +71,16 @@ const MeetingStateMachine = {
           ],
           to: MEETING_STATE_MACHINE.STATES.ENDED
         },
+        //  when ending a meeting it must be from  JOINED, or ERROR states, and transitions it to the ENDED state
```
[the remark this answers is not in the document]
Comment (sumitkumar121) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/state.js]: comment has been removed

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (marcin-bazyl) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:5635]: I'm confused - you say that it's confirmed that we need to send the same event for "leave" and "end", but the code here is sending `eventType.END` which is different to what `leave()` is doing

[the remark this answers is not in the document]
Comment (sumitkumar121) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: Thanks for pointing this out event name should be leave not end. Corrected this

File: packages/node_modules/@webex/plugin-meetings/src/meeting/util.js (hunk 4 above)
Comment (sumitkumar121) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/util.js]: cleaned the todo comment

Comment (sumitkumar121) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/util.js]: cleaned the todo comment

File: packages/node_modules/@webex/plugin-meetings/src/meeting/util.js (hunk 7 above)
Comment (marcin-bazyl) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/util.js:768]: I don't see it being fixed

Comment (sumitkumar121) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/util.js]: @marcin-bazyl This has bben fixed

File: packages/node_modules/@webex/plugin-meetings/src/meeting/state.js (hunk 3 above)
Comment (marcin-bazyl) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/state.js:76]: ok, thanks, maybe we should have a separate test for the case of 1-1 call, but I'm approving the PR anyway

**Review by marcin-bazyl — APPROVED** [review]

File: packages/node_modules/@webex/plugin-meetings/src/meeting/util.js (hunk 7 above)
Comment (marcin-bazyl) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/util.js]: ah, yes, sorry, was looking at the diff from not all of the commits

**Pull request merged**